Sources/Sentry/SentryBinaryImageCache.m

```diff
@@ -0,0 +1,121 @@
+#import "SentryBinaryImageCache.h"
+#import "SentryCrashBinaryImageCache.h"
+
+static void binaryImageWasAdded(const SentryCrashBinaryImage *image);
+
+static void binaryImageWasRemoved(const SentryCrashBinaryImage *image);
+
+@implementation BinaryImageInfo
+@end
+
+@interface
+SentryBinaryImageCache ()
+@property (nonatomic, strong) NSMutableArray<BinaryImageInfo *> *cache;
+- (void)binaryImageAdded:(const SentryCrashBinaryImage *)image;
+- (void)binaryImageRemoved:(const SentryCrashBinaryImage *)image;
+@end
+
+@implementation SentryBinaryImageCache
+
++ (SentryBinaryImageCache *)shared
+{
+    static SentryBinaryImageCache *instance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{ instance = [[self alloc] init]; });
+    return instance;
+}
+
+- (void)start
+{
+    _cache = [NSMutableArray array];
+    sentrycrashbic_registerAddedCallback(&binaryImageWasAdded);
+    sentrycrashbic_registerRemovedCallback(&binaryImageWasRemoved);
+}
+
+- (void)stop
+{
+    sentrycrashbic_registerAddedCallback(NULL);
+    sentrycrashbic_registerRemovedCallback(NULL);
+    _cache = nil;
+}
+
+- (void)binaryImageAdded:(const SentryCrashBinaryImage *)image
+{
+    BinaryImageInfo *newImage = [[BinaryImageInfo alloc] init];
+    newImage.name = [NSString stringWithCString:image->name encoding:NSUTF8StringEncoding];
+    newImage.address = image->address;
+    newImage.size = image->size;
+
+    @synchronized(self) {
+        NSUInteger left = 0;
+        NSUInteger right = _cache.count;
+
+        while (left < right) {
+            NSUInteger mid = (left + right) / 2;
+            BinaryImageInfo *compareImage = _cache[mid];
+            if (newImage.address < compareImage.address) {
+                right = mid;
+            } else {
+                left = mid + 1;
+            }
+        }
+
+        [_cache insertObject:newImage atIndex:left];
+    }
+}
+
+- (void)binaryImageRemoved:(const SentryCrashBinaryImage *)image
+{
+    @synchronized(self) {
+        NSInteger index = [self indexOfImage:image->address];
+        if (index >= 0) {
+            [_cache removeObjectAtIndex:index];
+        }
+    }
+}
+
+- (nullable BinaryImageInfo *)imageByAddress:(const uint64_t)address;
+{
+    @synchronized(self) {
+        NSInteger index = [self indexOfImage:address];
+        return index >= 0 ? _cache[index] : nil;
+    }
+}
+
+- (NSInteger)indexOfImage:(NSUInteger)address
```

Comment:
`m`: As we are going to call this for every frame, this is `O(n log m)` where n is the amount of frames and m the amount of images. Ideally, we would improve the search somehow so we can get this close to `O(n)`, but I have to think about how to solve this.

Reply:
I thought about it and think `O(n log m)` is acceptable. It's as fast as quicksort. That should be fine.